File: nim/2023/day1.nim
include ../prelude

let input = readInput(2023, 1)
let lines = input.strip.splitLines.map(s => s.strip)

proc digits(lines: seq[string]): int =
  result = 0
  for line in lines:
    var first = -1
    var last = -1

    for ch in line:
      if ch.isDigit.not:
        continue

      let v = ch.int - '0'.int

      if first == -1:
        first = v
      else:
        last = v

    result += first * 10 + (if last == -1: first else: last)

proc deliteralString(line: string): string =
  return line
    .replace("one", "o1e")
    .replace("two", "t2o")
    .replace("three", "th3ee")
    .replace("four", "f4r")
    .replace("five", "f5e")
    .replace("six", "s6x")
    .replace("seven", "s7n")
    .replace("eight", "e8t")
    .replace("nine", "n9e")

echo "P1: ", digits(lines)
echo "P2: ", digits(lines.map(deliteralString))

File: csharp/AoC.2023/Day1.cs
using AoC.Shared;
using AoC.Shared.Extensions;
using Xunit.Abstractions;

namespace AoC._2023;

public class Day1(ITestOutputHelper outputHelper) : SolutionBase(2023, 1, false, outputHelper)
{
    public override async Task<Answer> Part1(string input)
    {
        return input
            .SplitLines()
            .Select(line =>
                line.ToCharArray()
                    .Where(char.IsDigit)
                    .Select(c => int.Parse(c.ToString()))
                    .ToList())
            .Select(n => $"{n.First()}{n.Last()}")
            .Select(int.Parse)
            .Sum();
    }

    readonly Dictionary<string, string> _numberDict = new()
    {
        ["one"] = "1",
        ["two"] = "2",
        ["three"] = "3",
        ["four"] = "4",
        ["five"] = "5",
        ["six"] = "6",
        ["seven"] = "7",
        ["eight"] = "8",
        ["nine"] = "9",
    };

    private string ReplaceNums(string line)
    {
        for (var i = 0; i < line.Length + 1; i++)
        {
            for (var j = i + 1; j < line.Length + 1; j++)
            {
                if (_numberDict.TryGetValue(line[i..j], out var value))
                {
                    // Replaces twone as 2wone so we can match 2w1 later 
                    return line[..i] +
                           value +
                           ReplaceNums(line[(i + 1)..]);
                }
            }
        }

        return line;
    }

    public override async Task<Answer> Part2(string input)
    {
        return input
            .SplitLines()
            .Select(ReplaceNums)
            .Select(line =>
                line.ToCharArray()
                    .Where(char.IsDigit)
                    .Select(c => int.Parse(c.ToString()))
                    .ToList()).Select(n => $"{n.First()}{n.Last()}")
            .Select(int.Parse)
            .Sum();
    }
}